desktop-widgets/undocommands.cpp

```diff
@@ -200,47 +254,90 @@ void UndoRenumberDives::redo()
 	undo();
 }
 
-UndoRemoveDivesFromTrip::UndoRemoveDivesFromTrip(const QVector<dive *> &divesToRemoveIn) : divesToRemove(divesToRemoveIn)
+void UndoTripBase::redo()
 {
-	setText(gettextFromC::tr("remove dive(s) from trip"));
+	moveDivesBetweenTrips(divesToMove);
+
+	mark_divelist_changed(true);
+
+	// Finally, do the UI stuff:
+	MainWindow::instance()->refreshDisplay();
 }
 
-void UndoRemoveDivesFromTrip::undo()
+void UndoTripBase::undo()
 {
-	// first bring back the trip(s)
-	for (auto &trip: tripsToAdd) {
-		dive_trip *t = trip.release();	// Give up ownership
-		insert_trip(&t);		// Return ownership to backend
+	// Redo and undo do the same thing!
+	redo();
+}
+
+UndoRemoveDivesFromTrip::UndoRemoveDivesFromTrip(const QVector<dive *> &divesToRemove)
+{
+	setText(divesToRemove.size() == 1 ? gettextFromC::tr("remove dive from trip")
+					  : gettextFromC::tr("remove %1 dives from trip").arg(divesToRemove.size()));
+	divesToMove.divesToMove.reserve(divesToRemove.size());
+	for (dive *d: divesToRemove)
+		divesToMove.divesToMove.push_back( {d, nullptr} );
+}
+
+UndoRemoveAutogenTrips::UndoRemoveAutogenTrips()
+{
+	setText(gettextFromC::tr("remove autogenerated trips"));
+	// TODO: don't touch core-innards directly
+	int i;
+	struct dive *dive;
+	for_each_dive(i, dive) {
+		if (dive->divetrip && dive->divetrip->autogen)
+			divesToMove.divesToMove.push_back( {dive, nullptr} );
 	}
-	tripsToAdd.clear();
+}
 
-	for (auto &pair: divesToAdd)
-		add_dive_to_trip(pair.first, pair.second);
-	divesToAdd.clear();
-	mark_divelist_changed(true);
+UndoAddDivesToTrip::UndoAddDivesToTrip(const QVector<dive *> &divesToAddIn, dive_trip *trip)
+{
+	setText(divesToAddIn.size() == 1 ? gettextFromC::tr("add dives to trip")
+					 : gettextFromC::tr("add %1 dives to trip").arg(divesToAddIn.size()));
```

Comment:
that messes with the way plurals are handled in the translation. Please check http://doc.qt.io/qt-5/i18n-source-translation.html#handling-plurals
reading through other commits it's clear that you just followed the existing - bad - pattern.
It would be nice if you could fix it while you are touching the code - or we could do that later if that's distracting right now.

Reply:
this still needs to be addressed!